lua.go

```diff
@@ -0,0 +1,178 @@
+		"body_filter_by_lua_block",
+		"log_by_lua_block",
+		"balancer_by_lua_block",
+		"ssl_client_hello_by_lua_block",
+		"ssl_certificate_by_lua_block",
+		"ssl_session_fetch_by_lua_block",
+		"ssl_session_store_by_lua_block",
+	}
+}
+
+// RegisterLexer registers a lexer for parsing Lua blocks.
+func (l *Lua) RegisterLexer() RegisterLexer { //nolint:ireturn
+	return LexWithLexer(l, l.directiveNames()...)
+}
+
+// Lex lexically analyzes the Lua blocks based on directives detected.
+// It is used by the lexer to tokenize Lua content within configuration files.
+//
+//nolint:funlen,gocognit,gocyclo,nosec
+func (l *Lua) Lex(s *SubScanner, matchedToken string) <-chan NgxToken {
+	tokenCh := make(chan NgxToken)
+
+	tokenDepth := 0
+
+	go func() {
+		defer close(tokenCh)
+		var tok strings.Builder
+		var inQuotes bool
+
+		// special handling for'set_by_lua_block' directive
+		// ignore potential hardcoded credentials linter warning for "set_by_lua_block"
+		if matchedToken == "set_by_lua_block" /* #nosec G101 */ {
+			arg := ""
+			for {
+				if !s.Scan() {
+					return
+				}
+				next := s.Text()
+				if isSpace(next) {
+					if arg != "" {
+						tokenCh <- NgxToken{Value: arg, Line: s.Line(), IsQuoted: false}
+						break
+					}
+
+					for isSpace(next) {
+						if !s.Scan() {
+							return
+						}
+						next = s.Text()
+					}
+				}
+				arg += next
+			}
+		}
+
+		// check that Lua block starts correctly
+		for {
+			if !s.Scan() {
+				return
+			}
+			next := s.Text()
+
+			if !isSpace(next) {
+				if next != "{" {
+					lineno := s.Line()
+					tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: `expected "{" to start lua block`, Line: &lineno}}
+					return
+				}
+				tokenDepth++
+				break
+			}
+		}
+
+		// Grab everything in Lua block as a single token and watch for curly brace '{' in strings
+		for {
+			if !s.Scan() {
+				return
+			}
+
+			next := s.Text()
+			if err := s.Err(); err != nil {
+				lineno := s.Line()
+				tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: err.Error(), Line: &lineno}}
+			}
+
+			switch {
+			case next == "{" && !inQuotes:
+				tokenDepth++
+				if tokenDepth > 1 { // not the first open brace
+					tok.WriteString(next)
+				}
+
+			case next == "}" && !inQuotes:
+				tokenDepth--
+				if tokenDepth < 0 {
+					lineno := s.Line()
+					tokenCh <- NgxToken{Error: &ParseError{File: &lexerFile, What: `unexpected "}"`, Line: &lineno}}
+					return
+				}
+
+				if tokenDepth > 0 { // not the last close brace for it to be 0
+					tok.WriteString(next)
+				}
+
+				if tokenDepth == 0 {
+					tokenCh <- NgxToken{Value: tok.String(), Line: s.Line(), IsQuoted: true}
+					tokenCh <- NgxToken{Value: ";", Line: s.Line(), IsQuoted: false} // For an end to the Lua string based on the nginx bahavior
+					// See: https://github.com/nginxinc/crossplane/blob/master/crossplane/ext/lua.py#L122C25-L122C41
+					return
+				}
+
+			case next == `"` || next == "'":
+				inQuotes = !inQuotes
```

Comment:
how do we know the "closing quote" is of the same type? 
if i do the following, wouldn't this mess things up?
```
" ' foo ' "

```

Reply:
Good catch, thanks for pointing that out!  I added `var quoteType string` to track if quotes are matched